File: 3.3/stack_of_plates.py

class SetOfStacks(object):

    def __init__(self, top, threshold):
        """ Initialize with a top stack """
        self.top = top
        self.threshold = threshold

    def push(self, n):
        if self.top.size() >= self.threshold:
            # Create a new stack
            new_stack = Stack(n)
            new_stack.next = self.top
            self.top = new_stack
        else:
            self.top.push(n)

    def pop(self):
        if self.top.size() == 1:
            s = self.top
            self.top = self.top.next
            return s.pop()
        else:
            self.top.pop()

    def size(self):
        size = 0
        s = self.top
        while s is not None:
            size += 1
            s = s.next
        return size

class Stack(object):

    def __init__(self, top):
        """ Initialize with top stack node """
        self.top = top
        self.next = None # Point to the next stack of plates

    def pop(self):
        """ Remove the top element from the stack """
        top = self.top
        self.top = self.top.next
        return top

    def push(self, n):
        """ Add an element to the top of the stack """
        n.next = self.top
        self.top = n

    def isEmpty(self):
        """ Checks to see if the stack is empty """
        if self.top is None:
            return True
        else:
            return False

    def size(self):
        size = 0
        n = self.top
        while n is not None:
            size = size + 1
            n = n.next
        return size


class StackNode(object):

    def __init__(self, data):
        self.data = data
        self.next = None

File: 3.3/c++/stackofplates.cpp
#include "stackofplates.hpp"

void StackOfPlates::push(StackNode *sn) {
    if (this->top->size() == 3) {
        Stack *new_stack = new Stack();
        new_stack->next = this->top;
        this->top = new_stack;
    }
    this->top->push(sn);
}

bool StackOfPlates::isEmpty() {
    return (this->top == nullptr);
}

int StackOfPlates::size() {
    int size = 0;
    Stack *s = this->top;
    while (s != nullptr) {
        size += 1;
        s = s->next;
    }
    return size;
}

StackNode *StackOfPlates::pop() {
    StackNode *sn = this->top->pop();
    if (this->top->size() == 0) {
        this->top = this->top->next;
    }
    return sn;
}
